api/v1alpha1/gateway_types.go

```diff
@@ -165,30 +170,44 @@ type ListenerAddress struct {
 // selectors are associated with the Gateway.
 type RouteBindingSelector struct {
 	// NamespaceSelector specifies a set of namespace labels used for selecting
-	// routes to associate with the Gateway. If NamespaceSelector is defined,
-	// all routes in namespaces matching the NamespaceSelector are associated
-	// to the Gateway.
+	// routes to associate with the Gateway.
 	//
 	// An empty NamespaceSelector (default) indicates that routes from any
-	// namespace will be associated to this Gateway. This field is intentionally
-	// not a pointer because the nil behavior (no namespaces) is undesirable here.
+	// Namespace can be selected by this Gateway.
+	//
+	// The OnlyLocalRoutes field takes precedence over this field. This selector
+	// will only take effect when OnlyLocalRoutes is false.
 	//
 	// Support: Core
 	//
+	// +optional
 	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector" protobuf:"bytes,1,opt,name=namespaceSelector"`
+	// OnlyLocalRoutes is a boolean used to indicate if routes should only be
+	// selected from within the same namespace.
+	//
+	// When true (default) only routes in the same namespace as the Gateway
+	// will be selected.
+	//
+	// This field takes precedence over the NamespaceSelector field. That
+	// selector should only take effect when this field is set to false.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:default=true
+	// +optional
+	OnlyLocalRoutes bool `json:"onlyLocalRoutes" protobuf:"bytes,2,opt,name=onlyLocalRoutes"`
```

Comment:
Is there any concern around how "Local" may be interpreted in a multi-cluster context? The description is very clear, but It might be worth considering adding "Namespace" to the field name, something like `SameNamespaceOnly` to make it more self-descriptive.

Reply:
OnlySameNamespaceRoutes?

Reply:
Also, probably easier to use if it is:

AllowSameNamespace

and it's a union with set selected by the NamespaceSelector

Reply:
I don't think a union with a selector is what we want here. That would result in a default of all namespaces (`{}` and `true`) and the only way to restrict that would be to change the selector to something that would not match any namespaces. I think we actually want the opposite of that. We want to provide an easy way to limit routes to only the local namespace.

Reply:
I think we would still want to have NamespaceSelector be a pointer to align with the standard range of labelselector usage 

(nil => none, {} => all, ...}

As documented.

This way we don't have to have a different policy wrt to how the values are combined (one takes precedence if set etc)